pkg/util/tpm/tpm.go

```diff
@@ -0,0 +1,433 @@
+	t.tpmclient = tpmResourceClient
+
+	config.APIPath = "apis/tpm.coreos.com"
+	policyclient, err := dynamic.NewClient(&config)
+	if err != nil {
+		return err
+	}
+	policyresource := &unversioned.APIResource{
+		Kind:       "Policy",
+		Name:       "policies",
+		Namespaced: true,
+	}
+	policyresourceclient := policyclient.Resource(policyresource, "default")
+	t.PolicyClient = policyresourceclient
+
+	return nil
+}
+
+func (t *TPMHandler) GetPolicies() ([]map[string]PCRConfig, error) {
+	var configs []map[string]PCRConfig
+	var options api.ListOptions
+
+	unstructuredPolicies, err := t.PolicyClient.List(&options)
+	if err != nil {
+		return nil, err
+	}
+	for _, unstructuredPolicy := range unstructuredPolicies.Items {
+		config := make(map[string]PCRConfig)
+		policy := unstructuredPolicy.Object["policy"]
+		if policy == nil {
+			continue
+		}
+		policymap, ok := policy.(map[string]interface{})
+		if !ok {
+			glog.Errorf("Unable to decode unstructured object %s", unstructuredPolicy.GetName())
+			continue
+		}
+		for pcr, unstructuredpcrconfig := range policymap {
+			var pcrconfig PCRConfig
+			err = mapstructure.Decode(unstructuredpcrconfig, &pcrconfig)
+			if err != nil {
+				glog.Errorf("Unable to unmarshal policy json from %s", unstructuredPolicy.GetName())
+			}
+			pcrconfig.Source = unstructuredPolicy.GetName()
+			pcrconfig.Policyref = unstructuredPolicy.GetSelfLink()
+			config[pcr] = pcrconfig
+		}
+		configs = append(configs, config)
+	}
+	return configs, nil
+}
+
+func (t *TPMHandler) Get(address string, allowEmpty bool) (*Tpm, error) {
+	c := tpmclient.New(address, 30*time.Second)
+	ekcert, err := c.GetEKCert()
+
+	if err != nil {
+		return nil, err
+	}
+
+	eksha := sha1.Sum(ekcert)
+	ekhash := hex.EncodeToString(eksha[:])
+	tpm := &Tpm{}
+	unstructuredTpm, err := t.tpmclient.Get(ekhash)
+
+	if err != nil {
+		if allowEmpty == false {
+			return nil, fmt.Errorf("TPM does not exist and automatic creation is forbidden")
+		}
+		err = verification.VerifyEKCert(ekcert)
+		if err != nil {
+			return nil, err
+		}
+		tpm = &Tpm{
+			EKCert: ekcert,
+		}
+		unstructuredTpm.Object = make(map[string]interface{})
+		unstructuredTpm.SetKind("Tpm")
+		unstructuredTpm.SetAPIVersion("coreos.com/v1")
+		unstructuredTpm.SetName(ekhash)
+		unstructuredTpm.Object["EKCert"] = base64.StdEncoding.EncodeToString(tpm.EKCert)
+		unstructuredTpm.Object["AIKPub"] = ""
+		unstructuredTpm.Object["AIKBlob"] = ""
+
+		unstructuredTpm, err = t.tpmclient.Create(unstructuredTpm)
+		if err != nil {
+			return nil, err
+		}
+		unstructuredTpm, err = t.tpmclient.Get(ekhash)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tpm.EKCert, err = base64.StdEncoding.DecodeString(unstructuredTpm.Object["EKCert"].(string))
+	if err != nil {
+		glog.Errorf("Unable to decode TPM EK Cert from %s", unstructuredTpm.GetName())
+		return nil, err
+	}
+	tpm.AIKPub, _ = base64.StdEncoding.DecodeString(unstructuredTpm.Object["AIKPub"].(string))
```

Comment:
Is the error data potentially sensitive and why we don't need it?

Even so, we are returning err which really is from the EKCert decoding above, as its' not repopulated here